### PR TITLE
Disallow identifiers that begin with a double underscore.

### LINKED ISSUE #1533: Forbid double leading underscores in function, variable, struct, enum, etc. names.

### PR DESCRIPTION
**Pull request by sezna**
Closes #1533.

### CODE REVIEW

File: test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_fn/Forc.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "double_underscore_fn"
+
+[dependencies]
+core = { path = "../../../../../../sway-lib-core" }
```
Comment (adlerjohn) [inline test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_fn/Forc.toml:8]: If you're not using `core`, maybe use `implicit_std = false` instead?

File: test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_enum/src/main.sw [hunk 2]
```diff
@@ -0,0 +1,6 @@
+script;
+
+
```
Comment (adlerjohn) [inline test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_enum/src/main.sw:3]: fmt pls!

File: sway-parse/src/item/mod.rs [hunk 3]
```diff
@@ -278,6 +278,12 @@ impl Parse for FnSignature {
             }
             None => None,
         };
+        if name.as_str().starts_with("__") {
```
Comment (adlerjohn) [inline sway-parse/src/item/mod.rs:281]: HMMMM are you sure this doesn't conflict with #1536?

File: test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_enum/Forc.toml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "double_underscore_fn"
```
Comment (mohammadfawaz) [inline test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_enum/Forc.toml:5]:
```suggestion
name = "double_underscore_enum"
```

File: test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_var/Forc.toml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "double_underscore_fn"
```
Comment (mohammadfawaz) [inline test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_var/Forc.toml:5]:
```suggestion
name = "double_underscore_var"
```

File: test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_struct/Forc.toml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "double_underscore_fn"
```
Comment (mohammadfawaz) [inline test/src/e2e_vm_tests/test_programs/should_fail/double_underscore_struct/Forc.toml:5]:
```suggestion
name = "double_underscore_struct"
```

**Review by mohammadfawaz — CHANGES_REQUESTED** [review]

**Pull request closed without merging**